conn_writer.go

```diff
@@ -114,13 +105,27 @@ func (cw *connWriter) flush() bool {
 		if cw.write(mu) {
 			flushBuf = append(flushBuf, mu)
 		}
+
 	}
 	cw.flushBuf = cw.flushBuf[:0]
 
-	if err := cw.bw.Flush(); err != nil {
+	// we don't really honor the context deadline during the write phase, as it
+	// complicates things surrounding write+flush behavior, and the reader is
+	// going to check on it anyway. The write method does a basic check at the
+	// beginning to make sure that the context isn't already canceled, but
+	// that's it.
+	if err := cw.conn.SetWriteDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		err = fmt.Errorf("setting write deadline prior to flush: %w", err)
 		for _, mu := range flushBuf {
 			mu.errCh <- err
 		}
+
+	} else if err := cw.bw.Flush(); err != nil {
```

Comment:
If the Flush fails, should we not close the connection here to ensure that no subsequent request is written and forwarded to the reader?

Reply:
I'm open to that, but it would need to do so if the write fails as well. I think we should leave that out of this PR for now, and implement it alongside justin's suggestion of using something like github.com/oklog/run for go-routine management. The way go-routine management is done now would make doing this tricky, but with that package's pattern it would be a lot easier.